lib/Config/AccessControl.php

```diff
@@ -14,48 +14,41 @@ class AccessControl {
   const PERMISSION_MANAGE_SEGMENTS = 'mailpoet_manage_segments';
   const NO_ACCESS_RESTRICTION = 'mailpoet_no_access_restriction';
 
-  /** @var WPFunctions */
-  private $wp;
-
-  function __construct(WPFunctions $wp) {
-    $this->wp = $wp;
-  }
-
```

Comment:
Why do you remove DI from a class where it was already used? I thought that the singleton "hack" is something temporary like a stage in between transforming all classes to DI. What do we "downgrade" from that here?

Reply:
I didn't think about it as a hack, it makes mocking methods during tests easier. Let's take the following example:

```php
class A {
  function methodA() {
    //...
    B::methodB();
  }
}

class B {
  static function methodB() {
    $c = new C;
    $c->methodC();
    wp_call();
  }
}

class C {
  function methodC() {
    wp_call(...);
  }
}
```

If we want to use DI, we will need to:

- Add `$wp` as property of class `C` and inject it in the constructor.
- Add `$wp` as optional argument to `B::methodB` and do `new C($wp)`. 
- Add `$wp` as property of class `A` and inject it in the constructor. Even if class `A` doesn't call any WP function. But if we don't do that, we will not be able to mock WP calls when testing `A::methodA`.

This is just a small example, if we want to do that in our code, we will add `$wp` to the majority of our classes!

Reply:
Well, you could say that about every single class managed by DI and then we would not need DI at all :) It's very easy to use static service locator and just use it everywhere. But it's also not very good. It's also very easy to have everything global like WP does and that's even less good.

Adding a setter for tests (or anything else) is bad, using it has global side effects. In fact adding anything just for tests is bad, it means it is not a well-designed testable unit.

In your example I see many problems from my point of view - class `A` you should not use `B::methodB()` to initialize `B`, either the service or a non-static factory should be injected, `B` should not have a static factory method for `C`, if it is a factory, then normal method would do it, so it's hard to imagine such situation.